src/ktopt_planning_context.cpp

```diff
@@ -252,6 +253,73 @@ void KTOptPlanningContext::addPathPositionConstraints(KinematicTrajectoryOptimiz
   }
 }
 
+void KTOptPlanningContext::addPathOrientationConstraints(KinematicTrajectoryOptimization& trajopt,
+                                                         const MultibodyPlant<double>& plant,
+                                                         Context<double>& plant_context, const double padding)
+{
+  // retrieve the motion planning request
+  const auto& req = getMotionPlanRequest();
+
+  // check for path position constraints
+  if (!req.path_constraints.orientation_constraints.empty())
+  {
+    for (const auto& orientation_constraint : req.path_constraints.orientation_constraints)
+    {
+      // NOTE: Current thesis, when you apply a Drake Orientation Constraint it
+      // expects the following
+      // Frame A: The frame in which the orientation constraint is defined
+      // Frame B: The frame to which the orientation constraint is enforced
+      // theta_bound: The angle difference between frame A's orientation and
+      //              frame B's orientation
+
+      // Extract FrameA and FrameB for orientation constraint
+      // frame A
+      const auto link_ee_name = orientation_constraint.link_name;
+      if (!plant.HasBodyNamed(link_ee_name))
+      {
+        RCLCPP_ERROR(getLogger(),
+                     "The link specified in the PositionConstraint message, %s, does not exist in the Drake "
+                     "plant.",
+                     link_ee_name.c_str());
+        continue;
+      }
+      const auto& link_ee_frame = plant.GetFrameByName(link_ee_name);
+
+      // frame B
+      const auto base_frame_name = orientation_constraint.header.frame_id;
+      if (!plant.HasBodyNamed(base_frame_name))
+      {
+        RCLCPP_ERROR(getLogger(),
+                     "The link specified in the PositionConstraint message, %s, does not exist in the Drake "
+                     "plant.",
+                     base_frame_name.c_str());
+        continue;
+      }
+      const auto& base_frame = plant.GetFrameByName(base_frame_name);
+
+      // Extract the orientation of that Frame B needs to be constrained to w.r.t Frame A
+      const auto& constrained_orientation = orientation_constraint.orientation;
+      // convert to drake's RotationMatrix
+      const auto R_BbarB = RotationMatrixd(Eigen::Quaterniond(constrained_orientation.w, constrained_orientation.x,
+                                                              constrained_orientation.y, constrained_orientation.z));
+
+      // NOTE: There are 3 tolerances given for the orientation constraint in moveit
+      // message, rake only takes in a single theta bound. For now, we take any
+      // of the tolerances as the theta bound
+      const double theta_bound = orientation_constraint.absolute_x_axis_tolerance;
```

Comment:
`theta_bound` is not used!?

Reply:
Yeah, I think `theta_bound` and `padding` need to combined in some way in the constructor of `OrientationConstraint` below.

Probably just `theta_bound - padding`, or `theta_bound - 2.0 * padding` depending on how we want to enforce it?

Reply:
Also, since `theta_bound` uses the whole angle (per the [documentation](https://drake.mit.edu/doxygen_cxx/classdrake_1_1multibody_1_1_orientation_constraint.html#ab78d1f1f419685fc5ce820fcd7d8c4bb)), does it make sense to do something like take the norm of the X/Y/Z angle tolerances?

This is still not a proper calculation, so I think we anyhow have to be "creative" in how we interpret the tolerances in the MoveIt message vs. what translates to here.

Reply:
It is but only via the orientation padding parameter in the yaml file.

Reply:
I agree with https://github.com/moveit/moveit_drake/pull/55#discussion_r1800903157 to use the norm or sum of all angle tolerances. MoveIt's orientation constraint is much more flexible than Drake's one, allowing deviations to be specified for all rotation axes separately. It will only translate in a meaningful way if the MoveIt tolerances are all identical.

Reply:
So for this, a better approach would be to add an additional rotational transformation to `R_BbarB`, computing `R_BbarBx`, `R_BbarBy` and `R_BbarBz` - add 3 different orientation constraints for each axis with the theta_bounds being the individual tolerances.

Reply:
This will not work. Drake's orientation constraint already fully constrains all three axes. Adding more orientation constraints doesn't make sense.